src/kangaroo/solver.rs

```diff
@@ -0,0 +1,47 @@
+use crate::kangaroo::{self, Kangaroo};
+use crate::utils;
+
+use curve25519_dalek_ng::{constants::RISTRETTO_BASEPOINT_POINT, ristretto::RistrettoPoint};
+use std::ops::{Add, AddAssign, Mul, Sub};
+use std::time::{Duration, Instant};
+
+impl Kangaroo {
+    pub fn solve_dlp(&self, pk: &RistrettoPoint, max_time: Option<u64>) -> Option<u64> {
+        let start_time = max_time.map(|_| Instant::now());
```

Comment:
Since you're using map here to check whether max_time is provided, lets use this one:
```suggestion
        let start_time = max_time.and_then(|_| Some(Instant::now()));
```